Add files via upload

DemIS-1 opened the pull request:
File lock fix for Windows systems....
[coul.zip](https://github.com/hvds/divrep/files/10146960/coul.zip)

DemIS-1 committed:
Add files via upload

File lock fix for Windows systems....

hvds commented on coul.c:
```diff
@@ -752,6 +752,7 @@ void recover(void) {
             fail("unexpected log line %.3s in %s", curbuf, rpath);
     }
     fseek(rfp, 0L, SEEK_END);
+	fflush(rfp);
```
I don't think this should be needed, there is no unflushed write - the only possible modification up to now is the ftruncate(), which is a lower-level function.

DemIS-1 replied:
fflush()
"and the last data modification and last file status change timestamps of the underlying file shall be marked for update." See full description: [https://pubs.opengroup.org/onlinepubs/9699919799/functions/fflush.html](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fflush.html )

hvds replied:
The file modification time should reflect the time the file was last modified. What modification do you want to reflect here?

Note that ftruncate() also updates the modification timestamp.

DemIS-1 replied:
When we use fflush(rfp) (but not fflush()), we have two actions at once:
1. We clear what is possibly left in the stream buffer.
2. We give the go-ahead to change the timestamp.

hvds replied:
Yes, I understand what fflush does. However at this point there is nothing in the stream buffer, and there is no change that would cause us to want to change the timestamp.

That is why I asked "What modification do you want to reflect here?"

hvds commented on coul.c:
```diff
@@ -1057,6 +1058,7 @@ void report_init(FILE *fp, char *prog) {
     if (clock_is_realtime)
         fprintf(fp, " *RT*");
     fprintf(fp, "\n");
+	fflush(fp);
```
This makes sense, but identation is wrong - I suspect you have a tab there.

DemIS-1 replied:
[https://pubs.opengroup.org/onlinepubs/9699919799/functions/fflush.html](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fflush.html)

DemIS-1 replied:
"tab" - like symbol?
Didn't know it was a problem with your code.
Accepted.
I will fix this moment.

hvds replied:
Oh, I forgot that we have also called `setlinebuf()` on this filehandle, so it is automatically flushed when we write a newline to it. So again, this should not be needed.

`setlinebuf(rfp)` is equivalent to `setvbuf(rfp, NULL, 0, _IOLBF)`, see [setvbuf](https://pubs.opengroup.org/onlinepubs/9699919799/functions/setvbuf.html).

hvds commented on coul.c:
```diff
@@ -2916,7 +2918,13 @@ int main(int argc, char **argv, char **envp) {
 
     init_post();
     report_init(stdout, argv[0]);
+    if (freopen(rpath, "a+", rfp) == NULL)
+       fail("could not reopen %s: %s", rpath, strerror(errno));
+    setlinebuf(rfp);
```
1) For what purpose do you think this is needed?
2) These lines must be protected by an `if (rfp)` guard.
3) Please see line ~~273~~ 723 for the preferred form for calling freopen().
4) I'm not sure if the additional setlinebuf() call is needed: either it is unnecessary here, or it is also needed after line ~~273~~ 723.

DemIS-1 replied:
1. Reassign the file descriptor to the actual context point.
2. I think it's not necessary, because if there is no rfp object, then there will be an error message and there is an error handler, and if there is an object, then the buffer will work.
3. line 273 (272 ???) have other target (fflush(stdout)).

hvds replied:
1. I don't understand what you mean by that. Normally one would use freopen() only if a) you want to open a different path, or b) you want to open with a different mode. But neither of those are changing here, so it should be a null operation.
2. It is legitimate to run pcoul without a logfile - I do this all the time. When that is the case, rfp will be NULL and should not be written to.
3. Sorry, not sure where I got that line number from, it should be ~~720~~ 723: https://github.com/hvds/divrep/blob/1877d057621c934a90f37578af5226624b3eb1e8/coul.c#L723

hvds commented on coul.c:
```diff
@@ -2916,7 +2918,13 @@ int main(int argc, char **argv, char **envp) {
+       fail("could not reopen %s: %s", rpath, strerror(errno));
+    setlinebuf(rfp);
     if (rfp) report_init(rfp, argv[0]);
+    if (freopen(rpath, "a+", rfp) == NULL)
+       fail("could not reopen %s: %s", rpath, strerror(errno));
+    setlinebuf(rfp);
```
It makes no sense to me that this would be needed again here just because report_init() wrote a line.

DemIS-1 replied:
But we need to redefine (for Windows) after this function the file stream . But here we need to clarify this point with programmers who write programs for Windows.

hvds replied:
I don't understand, sorry. We have just called a function that will write a line to the output stream, this is a perfectly normal thing to do. Please try to explain in more detail why you think the file should be reopened after that (and not, for example, after every other time we write a line of logging to `rfp`).

hvds reviewed (commented):
Not sure why you described it as a "file lock fix", since no locking appears to occur. I added comments on the individual changes.

DemIS-1 committed:
Update coul.c

hvds commented:
I think it is worth stepping back a little, because I have not yet seen a clear description of what problem you are trying to fix here.

Please could you give a step-by-step description of how you reproduce the problem, eg:
* create a sample logfile with the following contents;
* show the current time using command <x>, show the current file modification time and filesize using command <y> giving the following results;
* in another window, start pcoul with the following commandline;
* wait <z> seconds, then show the current time and the current file modification time and filesize giving the following results.

To understand what is going on, it would also be useful to add a `printf()` message when the file is truncated (around coul.c:720). It may also be of interest to write a small program that is just enough to demonstrate the problem - open a file, read some data, write some data, sleep() so that you can check the file details. I can help with that if you wish.

DemIS-1 commented:
Log from begine b582:
```
07.12.2022 
06:38
07.12.2022  06:34                63 rep-b582.txt
==================
07.12.2022 
06:44
07.12.2022  06:44               143 rep-b582.txt
==================
07.12.2022 
06:46
07.12.2022  06:46               214 rep-b582.txt
==================
07.12.2022 
06:59
07.12.2022  06:56               306 rep-b582.txt
```

After restart b582 we see:
```
07.12.2022 
07:06
07.12.2022  07:06               387 rep-b582.txt
==================
07.12.2022 
07:10
07.12.2022  07:06               478 rep-b582.txt
==================
07.12.2022 
07:21
07.12.2022  07:06               566 rep-b582.txt
==================
```
The file time has stopped changing.
There is an important condition for seeing this: 
1. The program must be executed from a batch file.
2. The bat file must have a "for" type cycle.
3. The program must be executed from a "for" type loop.
For example:
```
for /L %%m in (1,1,10) do (
  echo pcoul -rlogs\rep-b%%m.txt -x:9887353188984012120346 -f11 -b%%m -p1e7 -g12 12 11
       pcoul -rlogs\rep-b%%m.txt -x:9887353188984012120346 -f11 -b%%m -p1e7 -g12 12 11
)
```
4. The program must be stopped with Ctrl+C and restarted with the same parameters.

DemIS-1 committed:
Redesigned call method to fix FMT on Windows

When the program is executed from a loop in a batch file, the FMT update stops.
The principle of opening a file in the program has been changed. To fix this glitch.

DemIS-1 committed:
Delete coul.c.zip

Wrong file

DemIS-1 committed:
Redesigned call method to fix FMT on Windows

When the program is executed from a loop in a batch file, the FMT update stops.
The principle of opening a file in the program has been changed. To fix this glitch.

Commit:
check and analyse mintau

DemIS-1 committed:
checked and analyze some line of code

hvds commented on coul.c:
```diff
@@ -259,7 +259,7 @@ void update_window(void) {
         /* update window title and icon with <ESC> ] 0 ; "string" <BEL> */
         uint this_batch = (opt_batch_min < 0) ? batch_alloc : batch_alloc - 1;
-        printf("\x1b]0;b%d:", this_batch);
+        printf("\x1b]0;b%d:", this_batch); // requires 'int' but the argument type is 'unsigned int'
```
Yes, this is a bug. (But it will go wrong only for a batch id >= 2^31.)

hvds commented on coul.c:
```diff
@@ -1053,11 +1053,11 @@ void report_init(FILE *fp, char *prog) {
     if (rough)
         fprintf(fp, " -h%u", rough);
     if (opt_batch_min >= 0) {
-        fprintf(fp, " -b%u", opt_batch_min);
+        fprintf(fp, " -b%u", opt_batch_min); // requires 'unsigned int' but the argument type is 'signed int'
```
It would be reasonable to change this, here and below, but note that it cannot actually go wrong since we are guarded by the check "opt_batch_min >= 0".

hvds commented:
I have edited your comment a bit: you can use a single backtick \`...\` only within a line. For multi-line blocks, you need \`\`\` on a line of its own before and after the block.

> Log from begine b582:
> 
> ```
> 07.12.2022 
> 06:38
> 07.12.2022  06:34                63 rep-b582.txt
> ```
What command generates this output? What do the various parts of the output mean?

> The file time has stopped changing. There is an important condition for seeing this:
> 
>     1. The program must be executed from a batch file.
> 
>     2. The bat file must have a "for" type cycle.
> 
>     3. The program must be executed from a "for" type loop.
>        For example:
> 
> 
> ```
> for /L %%m in (1,1,10) do (
>   echo pcoul -rlogs\rep-b%%m.txt -x:9887353188984012120346 -f11 -b%%m -p1e7 -g12 12 11
>        pcoul -rlogs\rep-b%%m.txt -x:9887353188984012120346 -f11 -b%%m -p1e7 -g12 12 11
> )
> ```

Presumably you would not actually use "(1,1,10)" in normal operation? How did the problem originally occur?

Note that I also recently discovered that, apparently, if a batch file is edited while Windows is running it, it will find the next command to execute by reopening the batch file and seeking to the spot it had previously read to. Thus if the file has been edited in preparation for the next run in a way that changes the length of a line, it may see a spurious command in the current run.

(I do not know if this is relevant for your case, but it is useful to know.)

DemIS-1 commented:
Thank you!
I didn't understand how it can be fixed. I'm sorry.

About command:
```
cat 1.bat
date /t >> reg-b582.log
time /t >> reg-b582.log
dir rep-b582.txt |grep 07.12.2022 >> reg-b582.log
```
(note that 'cat' and 'grep' is not a standard Windows program)

for example:
```
07.12.2022 
06:38
07.12.2022  06:34                63 rep-b582.txt
```
It is interpreted like this:
```
07.12.2022 <- Date when this script was run
06:38      <- Time when this script was run
```
the current status of the file on the system, where:
```
the date   | time  | size   |name file
07.12.2022 | 06:34 | 63     |rep-b582.txt
```
'dir rep-b582.txt' is like 'ls rep-b582.txt' on Unix.

1,1,10 is the iterator rule, very simple, where:
1 - What number to start with
1 - iteration step
10 - Which number to end on

The problem immediately arose (11/24/2022, when I did the first compilation and started the execution in the script  on Windows).
Only for a long time I could not understand that it was happening precisely because of the FOR.
I only understood yesterday.

I suppose that FOR does not understand all return signals.
I can be wrong.

It's like with the exit(1) operator, if it is outside the function main{}, then on windows it will always give us an exception error. Many compilers for Windows immediately show during compilation that this is a code error and the assembly does not occur (and some know how to fix on the fly).

DemIS-1 committed:
Merge branch 'master' of github.com:DemIS-1/divrep
Update branch

DemIS-1 committed:
Removed some of my comments

hvds commented:
The pull request was not being shown because you closed it. I am reopening it, all the information is still here.

DemIS-1 committed:
Merge branch 'master' of https://github.com/DemIS-1/divrep
